### PR TITLE
better reader + ArrayList

### PR DESCRIPTION
**Pull request by jgiron42**
add debug.assert

UPDATE Allocator:
add alignedAlloc

ADD array_list:
add most of the ArrayList and ArrayListAligned types

UPDATE Reader:
- use AnyReader rather than GenericReader,
- implement most of the type AnyReader
- update the shell to use readUntilDelimiterAlloc rather than just read

### CODE REVIEW

File: src/shell/token.zig [hunk 1]
```diff
@@ -69,11 +69,11 @@ pub fn tokenize(data: *[]u8) TokenizerError![][]u8 {
...
                 ft.mem.copyForwards(
                     u8,
-                    data.*[tokenizer.quote_offset .. i - 1],
-                    data.*[tokenizer.quote_offset + 1 .. i],
+                    data[tokenizer.quote_offset .. i - 1],
+                    data[tokenizer.quote_offset + 1 .. i],
                 );
-                ft.mem.copyForwards(u8, data.*[i - 1 .. data.*.len - 1], data.*[i + 1 .. data.*.len]);
-                @memset(data.*[data.*.len - 2 .. data.*.len], 0);
+                ft.mem.copyForwards(u8, data[i - 1 .. data.len - 1], data[i + 1 .. data.len]);
```
Comment (riblanc) [inline src/shell/token.zig:75]:
```suggestion
                @memcpy(data[i - 1 .. data.len - 1], data[i + 1 .. data.len]);
```
Idk why it did'nt worked the first time we tried but it works now, trust me :eyes:

File: src/shell.zig
Comment (riblanc) [inline src/shell.zig:1]: Such a pleasure to finally see my allocator in action!

File: src/ft/io/Reader.zig [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+// pub fn readBoundedBytes(self: Self, comptime num_bytes: usize) anyerror!std.BoundedArray(u8, num_bytes)
+
+pub fn readByte(self: Self) anyerror!u8 {
+    var buffer: [1]u8 = undefined;
+    try self.readNoEof(buffer[0..]);
+    return buffer[0];
+}
+
+pub fn readByteSigned(self: Self) anyerror!i8 {
+    return @bitCast(self.readByte());
```
Comment (riblanc) [inline src/ft/io/Reader.zig:94]:
```suggestion
    return @bitCast(try self.readByte());
```
I'm not 100% sure, but i feel it makes more sens to `try` to readByte to properly return the error if an error occurs during readByte call, doesn't it?

File: src/ft/array_list.zig [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+        }
+
+        pub fn clone(self: Self) Allocator.Error!Self {
+            const ret = Self{
+                .allocator = self.allocator,
+                .slice = try self.allocator.alignedAlloc(T, alignment, self.capacity),
+                .capacity = self.capacity,
+            };
+            ret.slice.len = self.slice.len;
+            @memcpy(ret.slice, self.slice);
```
Comment (riblanc) [inline src/ft/array_list.zig:161]:
```suggestion
            @memcpy(ret.slice, self.slice);
            return ret;
```

**Review by riblanc — CHANGES_REQUESTED** [review]

File: src/shell/token.zig (hunk 1 above)
Comment (jgiron42) [reply src/shell/token.zig:75]: Memcpy is undefined with overlapping regions

Comment (riblanc) [reply src/shell/token.zig:75]: Oooooh yip

**Review by riblanc — APPROVED** [review]

**Pull request merged**